### DISCUSSION TITLE
UseInteractions return types

### DISCUSSION THREAD

**Discussion by james-Infotrack** [floating-ui/floating-ui#2266] [React]
**Is your feature request related to a problem? Please describe.**
Hey everyone, I love your library,  I'm making a POC to migrate all our dropdown components over to it.

Just having trouble really understanding the types and properties that `getReferenceProps` and `getFloatingProps` returns when we are spreading them into the components. 

```ts
   getReferenceProps: (userProps?: React.HTMLProps<Element>) => Record<string, unknown>;
    getFloatingProps: (userProps?: React.HTMLProps<HTMLElement>) => Record<string, unknown>;
```
Their ts definitions are very vague.

**Describe the solution you'd like**
useInteractions added to the documentation website, clearly outlining what the return types are.

**Describe alternatives you've considered**
I tried reading the code and types inside 
https://github.com/floating-ui/floating-ui/blob/master/packages/react/src/useInteractions.ts

But it didn't offer a lot of clarity 
my understanding so far is that it passes in what ever props we spread and then some elementprops 
```ts
export interface ElementProps {
  reference?: React.HTMLProps<Element>;
  floating?: React.HTMLProps<HTMLElement>;
  item?: React.HTMLProps<HTMLElement>;
}
```
But I was under the assumption it also passes in some aria-attributes which I cannot find in the types. 

Any help with this would be greatly appreciated

**Comment by atomiks** [comment]
An [interaction hook](https://floating-ui.com/docs/react#interaction-hooks) returns an object of 3 (optional) properties, where each is an object of HTML props:

```js
{
  reference: {}
  floating: {},
  item: {},
}
```

```js
function useOne(context: FloatingContext) {
  return React.useMemo(() => ({
    reference: {
      onClick: () => console.log('clicked one'),
    }
  }), []);
}

function useTwo(context: FloatingContext) {
  return React.useMemo(() => ({
    reference: {
      onClick: () => console.log('clicked two'),
    }
  }), []);
}
```

`useInteractions` merges those objects into prop getters, so interaction hooks can each specify an event listener multiple times and they get composed together without overwriting eachother: 

```js
const {getReferenceProps} = useInteractions([useOne(), useTwo()]);

getReferenceProps();

// returns
{
  // useOne and useTwo's listener get merged together:
  onClick() {
    console.log('clicked one');
    console.log('clicked two');
  }
}
```

You can pass props into the prop getters to add your own event listener at call time, without creating a [custom hook](https://floating-ui.com/docs/custom-hooks):

```js
getReferenceProps({
  onClick() {
    console.log('clicked three');
  }
});

// returns
{
  onClick() {
    console.log('clicked one');
    console.log('clicked two');
    console.log('clicked three');
  }
}
```

The props don't have to be event listeners though, some of the interaction hooks add `aria-activedescendant`, etc. These are part of the object that calling the prop getter returns.

**Reply by james-Infotrack** [reply]
Thanks for replying!, 

Seems I may have misunderstood a bit, so on it's own the Interactions don't actually add any props, it's only the result of merging all the hooks together. 


```ts
 const { getReferenceProps, getFloatingProps, getItemProps } = useInteractions([
    click,
    dismiss,
    role,
    listNav,
    typeahead,
  ]);
```

here for example, I'm building a <Select /> component and I've merged a few of these all together.   

but had the above example just been 

```ts
 const { getReferenceProps, getFloatingProps, getItemProps } = useInteractions();
```
basically my props would be empty? 

Am I understanding that correctly? Also thankyou for pointing out the link to them in the docs.! 

Cheers

**Reply by atomiks** [reply]
`useInteractions` adds `tabIndex: -1` to `getFloatingProps()` by itself, but the others would be empty yes. Some of them register effects instead of returning props though, which rely on the `enabled` prop instead (if avoiding passing it to the array).
